"""
üöÄ COMPREHENSIVE DEMO - Distributed Key-Value Store
Demonstrates all phases: Distribution, Replication, Failure Detection, Recovery
"""

import sys
import os
import grpc
import time
import subprocess

# Fix import path
project_root = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, project_root)

from src.proto import kvstore_pb2
from src.proto import kvstore_pb2_grpc


# ============================================================================
# Helper Functions
# ============================================================================

def print_header(title, symbol="="):
    """Print formatted header."""
    print(f"\n{symbol * 80}")
    print(f"  {title}")
    print(f"{symbol * 80}\n")


def connect_to_node(port):
    """Connect to a node."""
    address = f'localhost:{port}'
    channel = grpc.insecure_channel(address)
    stub = kvstore_pb2_grpc.KeyValueStoreStub(channel)
    return channel, stub


def list_keys_on_node(port):
    """Get all keys on a node."""
    try:
        channel, stub = connect_to_node(port)
        request = kvstore_pb2.ListKeysRequest()
        response = stub.ListKeys(request, timeout=5.0)
        channel.close()
        return list(response.keys) if response.keys else []
    except Exception as e:
        return []


def check_nodes_running(nodes):
    """Check if nodes are running."""
    running = {}
    for port, node_id in nodes:
        try:
            channel, stub = connect_to_node(port)
            request = kvstore_pb2.ListKeysRequest()
            stub.ListKeys(request, timeout=2.0)
            channel.close()
            running[node_id] = True
        except:
            running[node_id] = False
    return running


# ============================================================================
# Demo Phases
# ============================================================================

def demo_phase1_basic_operations():
    """Phase 1 & 2: Basic PUT/GET/DELETE operations."""
    print_header("PHASE 1 & 2: Basic Operations (Single Node)", "=")
    
    print("Testing basic operations on node1...")
    
    channel, stub = connect_to_node(8001)
    
    # PUT
    print("\n[1] Testing PUT operation:")
    request = kvstore_pb2.PutRequest(key="user:alice", value="Alice Smith")
    response = stub.Put(request)
    if response.success:
        print(f"   ‚úÖ PUT successful: user:alice = 'Alice Smith'")
        print(f"      Node: {response.node_id}, Replicas: {response.replicas_count}")
    
    # GET
    print("\n[2] Testing GET operation:")
    request = kvstore_pb2.GetRequest(key="user:alice")
    response = stub.Get(request)
    if response.found:
        print(f"   ‚úÖ GET successful: user:alice = '{response.value}'")
        print(f"      Node: {response.node_id}")
    
    # DELETE
    print("\n[3] Testing DELETE operation:")
    request = kvstore_pb2.DeleteRequest(key="user:alice")
    response = stub.Delete(request)
    if response.success:
        print(f"   ‚úÖ DELETE successful: user:alice")
    
    # Verify deleted
    request = kvstore_pb2.GetRequest(key="user:alice")
    response = stub.Get(request)
    if not response.found:
        print(f"   ‚úÖ Verified: Key not found after delete")
    
    channel.close()


def demo_phase3_distribution():
    """Phase 3: Consistent Hashing & Data Distribution."""
    print_header("PHASE 3: Consistent Hashing & Distribution", "=")
    
    print("Inserting 30 keys across cluster...")
    
    channel, stub = connect_to_node(8001)
    
    # Insert keys
    for i in range(30):
        key = f"data_{i}"
        value = f"value_{i}"
        request = kvstore_pb2.PutRequest(key=key, value=value)
        stub.Put(request, timeout=5.0)
        if (i + 1) % 10 == 0:
            print(f"   Progress: {i + 1}/30 keys inserted")
    
    channel.close()
    time.sleep(1)
    
    # Check distribution
    print("\n[Distribution Analysis]")
    nodes = [(8001, "node1"), (8002, "node2"), (8003, "node3")]
    
    for port, node_id in nodes:
        keys = list_keys_on_node(port)
        percentage = (len(keys) / 30) * 100
        print(f"   {node_id}: {len(keys)} keys ({percentage:.1f}%)")
    
    print("\n   ‚úÖ Data distributed via consistent hashing")


def demo_phase4_replication():
    """Phase 4: Replication (2 copies per key)."""
    print_header("PHASE 4: Replication", "=")
    
    print("Testing replication (each key stored on 2 nodes)...")
    
    channel, stub = connect_to_node(8001)
    
    # Insert a test key
    test_key = "replicated:test"
    print(f"\n[1] Inserting key: {test_key}")
    request = kvstore_pb2.PutRequest(key=test_key, value="Replicated Value")
    response = stub.Put(request, timeout=5.0)
    
    if response.success:
        print(f"   ‚úÖ PUT successful")
        print(f"      Primary node: {response.node_id}")
        print(f"      Total replicas: {response.replicas_count}")
    
    channel.close()
    time.sleep(1)
    
    # Check which nodes have this key
    print(f"\n[2] Checking replication across nodes:")
    nodes = [(8001, "node1"), (8002, "node2"), (8003, "node3")]
    nodes_with_key = []
    
    for port, node_id in nodes:
        keys = list_keys_on_node(port)
        if test_key in keys:
            nodes_with_key.append(node_id)
            print(f"   ‚úÖ {node_id}: Has key '{test_key}'")
        else:
            print(f"   ‚ö™ {node_id}: No key")
    
    if len(nodes_with_key) >= 2:
        print(f"\n   ‚úÖ Replication verified: Key exists on {len(nodes_with_key)} nodes")
    else:
        print(f"\n   ‚ö†Ô∏è  Only {len(nodes_with_key)} copies found")


def demo_phase5_forwarding():
    """Phase 3 & 4: Request Forwarding."""
    print_header("PHASE 3: Request Forwarding", "=")
    
    print("Testing: Client connects to ANY node, requests forwarded to owner...")
    
    test_key = "forward:test"
    
    # PUT from node1
    print(f"\n[1] PUT {test_key} via node1:")
    channel, stub = connect_to_node(8001)
    request = kvstore_pb2.PutRequest(key=test_key, value="Forward Test")
    response = stub.Put(request)
    actual_node = response.node_id
    print(f"   ‚úÖ Data stored on: {actual_node}")
    channel.close()
    
    # GET from all nodes
    print(f"\n[2] GET {test_key} from each node:")
    nodes = [(8001, "node1"), (8002, "node2"), (8003, "node3")]
    
    for port, node_id in nodes:
        try:
            channel, stub = connect_to_node(port)
            request = kvstore_pb2.GetRequest(key=test_key)
            response = stub.Get(request, timeout=5.0)
            channel.close()
            
            if response.found:
                print(f"   ‚úÖ {node_id}: Found '{test_key}' = '{response.value}'")
            else:
                print(f"   ‚ùå {node_id}: Not found")
        except Exception as e:
            print(f"   ‚ùå {node_id}: Error - {e}")
    
    print("\n   ‚úÖ Request forwarding works correctly")


def demo_phase6_failure_simulation():
    """Phase 5: Failure Detection (simulated)."""
    print_header("PHASE 5: Failure Detection", "=")
    
    print("Heartbeat and failure detection is running in background.")
    print("To test:")
    print("   1. Kill one node (Ctrl+C in its terminal)")
    print("   2. Wait 15 seconds")
    print("   3. Other nodes will detect it as failed")
    print("   4. Check server logs for 'Node X detected as FAILED' message")
    print("\n   ‚ÑπÔ∏è  Heartbeat interval: 5 seconds")
    print("   ‚ÑπÔ∏è  Failure timeout: 15 seconds")
    print("\n   ‚úÖ Failure detection is active")


def demo_phase7_recovery():
    """Phase 6: Data Recovery via GetSnapshot."""
    print_header("PHASE 6: Data Recovery (GetSnapshot)", "=")
    
    print("Testing GetSnapshot for recovery...")
    
    # Request snapshot from node1
    print("\n[1] Requesting snapshot from node1:")
    
    try:
        channel = grpc.insecure_channel('localhost:8001')
        stub = kvstore_pb2_grpc.NodeServiceStub(channel)
        
        request = kvstore_pb2.SnapshotRequest()
        response = stub.GetSnapshot(request, timeout=5.0)
        channel.close()
        
        print(f"   ‚úÖ Snapshot received")
        print(f"      Provider node: {response.provider_node_id}")
        print(f"      Total keys: {response.total_keys}")
        print(f"      Timestamp: {response.snapshot_timestamp}")
        
        if len(response.data) > 0:
            print(f"\n   Sample data (first 5):")
            for i, kv in enumerate(response.data[:5]):
                print(f"      {i+1}. {kv.key} = {kv.value}")
        
        print("\n   ‚úÖ Recovery mechanism ready")
        print("   ‚ÑπÔ∏è  A recovering node can call GetSnapshot to restore data")
    
    except Exception as e:
        print(f"   ‚ùå Error: {e}")


# ============================================================================
# Main Demo Flow
# ============================================================================

def main():
    """Run comprehensive demo."""
    
    print("\n" + "üöÄ" * 40)
    print("  DISTRIBUTED KEY-VALUE STORE - FULL SYSTEM DEMO")
    print("üöÄ" * 40)
    
    # Check nodes
    print("\n[Pre-check] Verifying cluster nodes...")
    nodes = [(8001, "node1"), (8002, "node2"), (8003, "node3")]
    running = check_nodes_running(nodes)
    
    for node_id, is_running in running.items():
        status = "‚úÖ Running" if is_running else "‚ùå Not Running"
        print(f"   {node_id}: {status}")
    
    if not all(running.values()):
        print("\n‚ùå Not all nodes are running!")
        print("Please start the cluster first:")
        print("   Option 1: Start background jobs:")
        print("   Start-Job -ScriptBlock { python F:\\app\\distributed-system\\distributed-kvstore\\src\\server.py 8001 node1 }")
        print("   Start-Job -ScriptBlock { python F:\\app\\distributed-system\\distributed-kvstore\\src\\server.py 8002 node2 }")
        print("   Start-Job -ScriptBlock { python F:\\app\\distributed-system\\distributed-kvstore\\src\\server.py 8003 node3 }")
        print("\n   Option 2: Start in separate terminals")
        return
    
    print("\n‚úÖ All nodes running. Starting demo...\n")
    time.sleep(2)
    
    # Run demos
    try:
        demo_phase1_basic_operations()
        time.sleep(2)
        
        demo_phase3_distribution()
        time.sleep(2)
        
        demo_phase4_replication()
        time.sleep(2)
        
        demo_phase5_forwarding()
        time.sleep(2)
        
        demo_phase6_failure_simulation()
        time.sleep(2)
        
        demo_phase7_recovery()
        
    except Exception as e:
        print(f"\n‚ùå Demo error: {e}")
        import traceback
        traceback.print_exc()
    
    # Final summary
    print_header("DEMO COMPLETE - Summary", "=")
    print("‚úÖ Phase 1 & 2: Basic operations work")
    print("‚úÖ Phase 3: Consistent hashing distributes data")
    print("‚úÖ Phase 3: Request forwarding works")
    print("‚úÖ Phase 4: Replication creates 2 copies")
    print("‚úÖ Phase 5: Failure detection active (background)")
    print("‚úÖ Phase 6: Recovery via GetSnapshot ready")
    print("\nüéâ All core features demonstrated successfully!")
    print("\nNext steps:")
    print("   - Test failure scenarios (kill a node)")
    print("   - Monitor logs in server.log")
    print("   - Try interactive client: python src/client.py")
    print("\n" + "=" * 80 + "\n")


if __name__ == '__main__':
    main()
